### PR TITLE
Session & Credentials Gauges

### PR DESCRIPTION
**Pull request by eliska-n**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,11 +24,8 @@
 
 ### Features
 - Cookie authentication in multi-domain setting (!219, PLUM Sprint 211217)
-- Endpoints for single tenant un/assignment (#2, PLUM Sprint 210114)
-- Endpoints for single role un/assignment (#3, PLUM Sprint 210114)
-
-### Refactoring
-- Remove `authz:credentials:admin` resource (#5, PLUM Sprint 210114)
+- Endpoints for simple tenant un/assignment (#2, PLUM Sprint 210114)
+- Login success/fail counters (#6, PLUM Sprint 220114)
```
Comment (byewokko) [inline CHANGELOG.md:28]: This looks like an incorrectly resolved rebase. You are accidentally overwriting new changes with old lines:)

**Review by byewokko — APPROVED** [review]
Looks good!

Just please resolve that minor changelog thing before merging.

File: seacatauth/credentials/service.py [hunk 2]
```diff
@@ -86,6 +86,25 @@ def __init__(self, app, service_name='seacatauth.CredentialsService', tenant_ser
...
+	async def _on_tick_metric(self, event_name):
+		tasks = []
+		for order, provider in self.Providers:
+			tasks.append(provider.count())
+		counts = await asyncio.gather(*tasks)
+		total = 0
+		for count in counts:
+			if count != -1:
+				total += count
+		self.CredentialsGauge.set("credentials", total)
```
Comment (ateska) [inline seacatauth/credentials/service.py:105]: As discussed, this should be done in the ASAB Task Service to off-load that from "on tick".

**Review by ateska — CHANGES_REQUESTED** [review]
One request, then mergeable.

File: seacatauth/authn/service.py [hunk 3]
```diff
@@ -81,8 +81,8 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
 		self.LoginSessions = {}
 
 		# Metrics - login counters
-		self.LoginFailedCounter = self.MetricsService.create_counter("login_failed", tags={"help": "Counts failed logins.", "unit": "logins", "default_label":"SeaCatAuth Metrics"}, init_values={"logins": 0})
-		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins.", "unit": "logins", "default_label":"SeaCatAuth Metrics"}, init_values={"logins": 0})
+		self.LoginFailedCounter = self.MetricsService.create_counter("login_failed", tags={"help": "Counts failed logins.", "unit": "logins", "default_label": "SeaCatAuth_Metrics"}, init_values={"logins": 0})
+		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins.", "unit": "logins", "default_label": "SeaCatAuth_Metrics"}, init_values={"logins": 0})
```
Comment (ateska) [inline seacatauth/authn/service.py:85]: Is the label specific enough?

File: seacatauth/session/service.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ def __init__(self, app, service_name='seacatauth.SessionService'):
 
 		# Metrics
 		self.MetricsService = app.get_service('asab.MetricsService')
-		self.SessionGauge = self.MetricsService.create_gauge("active_sessions", tags={"help": "Counts active sessions.", "unit": "sessions", "default_label":"SeaCatAuth Metrics"}, init_values={"sessions": 0})
+		self.SessionGauge = self.MetricsService.create_gauge("active_sessions", tags={"help": "Counts active sessions.", "unit": "sessions", "default_label": "SeaCatAuth_Metrics"}, init_values={"sessions": 0})
```
Comment (ateska) [inline seacatauth/session/service.py:75]: dtto

**Review by ateska — CHANGES_REQUESTED** [review]

File: seacatauth/authn/service.py (hunk 3 above)
Comment (eliska-n) [reply seacatauth/authn/service.py:85]: Not at all. It's one of my many questions. I'd appreciate a broader discussion.

Comment (ateska) [reply seacatauth/authn/service.py:85]: Ok - can you provide a sample of the output then.

Comment (eliska-n) [reply seacatauth/authn/service.py:85]: It sucks. But naming is just part of where I am stuck right now and the last outcome of several more questions... 
```
# TYPE active_sessions_sessions_sessions gauge
# UNIT active_sessions_sessions_sessions sessions
# HELP active_sessions_sessions_sessions Counts active sessions.
active_sessions_sessions_sessions{default_label="SeaCatAuth_Metrics"} 2
# TYPE provider1_credentials_users gauge
# UNIT provider1_credentials_users users
# HELP provider1_credentials_users Counts credentials.
provider1_credentials_users{default_label="SeaCatAuth_Metrics"} 4
# TYPE provider2_credentials_users gauge
# UNIT provider2_credentials_users users
# HELP provider2_credentials_users Counts credentials.
provider2_credentials_users{default_label="SeaCatAuth_Metrics"} 0
# TYPE login_failed_logins_logins counter
# UNIT login_failed_logins_logins logins
# HELP login_failed_logins_logins Counts failed logins.
login_failed_logins_logins_total{default_label="SeaCatAuth_Metrics"} 8
# TYPE login_success_logins_logins counter
# UNIT login_success_logins_logins logins
# HELP login_success_logins_logins Counts successful logins.
login_success_logins_logins_total{default_label="SeaCatAuth_Metrics"} 0
# EOF
```

**Comment by eliska-n** [comment]
I changed how metrics are intialized based on asab rearrangements. 
I added metrics_middleware to WebContainer. However, as this middleware produces only one Counter, starting the middleware in PublicWebContainer would rewrite the first one. Thus, it is implemented for one container ony.

File: seacatauth/authn/service.py [hunk 5]
```diff
@@ -81,8 +81,11 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
 		self.LoginSessions = {}
 
 		# Metrics - login counters
-		self.LoginFailedCounter = self.MetricsService.create_counter("login_failed", tags={"help": "Counts failed logins."}, init_values={"logins": 0})
-		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins."}, init_values={"logins": 0})
+		self.LoginCounter = self.MetricsService.create_counter(
+			"sca_login",
```
Comment (ateska) [inline seacatauth/authn/service.py:85]: `login` is ok, "sca_" prefix is not, we don't use this anywhere officially ... it is either `seacat_auth_login` or just `login`.

File: seacatauth/authn/service.py [hunk 6]
```diff
@@ -81,8 +81,11 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
...
+			tags={"help": "Counts successful and failed logins per minute.", "unit": "logins_per_minute"},
```
Comment (ateska) [inline seacatauth/authn/service.py:86]: `events_per_minute` that more universal unit.

File: seacatauth/authn/service.py [hunk 7]
```diff
@@ -81,8 +81,11 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
...
+			init_values={"successful_logins": 0, "failed_logins": 0}
```
Comment (ateska) [inline seacatauth/authn/service.py:87]: Why not `successful` / `failed` ?

File: seacatauth/credentials/service.py [hunk 8]
```diff
@@ -86,6 +86,31 @@ def __init__(self, app, service_name='seacatauth.CredentialsService', tenant_ser
 			self.register(provider)
 
 
+		# Metrics
+		self.MetricsService = app.get_service('asab.MetricsService')
+		self.TaskService = app.get_service('asab.TaskService')
+		self.Providers = providers
+		self.CredentialsGauge = self.MetricsService.create_gauge(
+			"sca_credentials",
```
Comment (ateska) [inline seacatauth/credentials/service.py:94]: No prefix, please.

File: seacatauth/session/service.py [hunk 9]
```diff
@@ -70,6 +70,12 @@ def __init__(self, app, service_name='seacatauth.SessionService'):
 		app.PubSub.subscribe("Application.tick/60!", self._on_tick)
 		app.PubSub.subscribe("Application.run!", self._on_start)
 
+		# Metrics
+		self.MetricsService = app.get_service('asab.MetricsService')
+		self.TaskService = app.get_service('asab.TaskService')
+		self.SessionGauge = self.MetricsService.create_gauge("sca_active_sessions", tags={"help": "Counts active sessions.", "unit": "sessions"}, init_values={"active_sessions": 0})
```
Comment (ateska) [inline seacatauth/session/service.py:76]: `sessions` ... "active" is implicit, "sca_" is out.

**Review by ateska — CHANGES_REQUESTED** [review]

**Review by byewokko — APPROVED** [review]
:unicorn: :ok_hand:

**Comment by eliska-n** [comment]
These changes are strongly dependent on https://github.com/TeskaLabs/asab/pull/197

**Review by ateska — APPROVED** [review]

**Pull request merged**